### PR TITLE
[SPARK-41759][CORE] Use `weakIntern` on string values in create new objects during deserialization

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to use weakIntern on string values in create new objects during deserialization.

### Why are the changes needed?
Following guid: https://github.com/apache/spark/pull/39270.
image

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Pass GA.

### CODE REVIEW

**Comment by mridulm** [comment]
Primarily use weakIntern for cases where there are a large number of duplicated strings with same value (so app start won't qualify), for the most common values.
Not for others

**Comment by LuciferYang** [comment]
@panbingkun We may need to investigate whether this change will change GC behavior

**Comment by LuciferYang** [comment]
@gengliangwang @mridulm @panbingkun 

Should we continue or close this one? I think we should complete this ticket  before  cut branch-3.4, , no matter how.

**Comment by gengliangwang** [comment]
I think we can use the weakIntern where the method is used before this new project. 
@panbingkun can you resolve conflicts and finish this one?

**Comment by panbingkun** [comment]
> I think we can use the weakIntern where the method is used before this new project.

Ok, let me do it.

**Comment by panbingkun** [comment]
> I think we can use the weakIntern where the method is used before this new project. 
> 
> @panbingkun can you resolve conflicts and finish this one?

Done

File: core/src/main/scala/org/apache/spark/status/protobuf/JobDataWrapperSerializer.scala [hunk 1]
```diff
@@ -79,16 +80,16 @@ class JobDataWrapperSerializer extends ProtobufSerDe[JobDataWrapper] {
   }
 
   private def deserializeJobData(info: StoreTypes.JobData): JobData = {
-    val description = getOptional(info.hasDescription, info.getDescription)
+    val description = getOptional(info.hasDescription, () => weakIntern(info.getDescription))
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/status/protobuf/JobDataWrapperSerializer.scala:83]: I wonder if it's worthwhile to intern strings that aren't likely to be repeated. Descriptions don't seem to be that case. IDs and names also seem more unique-ish. We might want to be more targeted? there is overhead to interning

Comment (gengliangwang) [reply core/src/main/scala/org/apache/spark/status/protobuf/JobDataWrapperSerializer.scala:83]: +1, I don't think we need weakIntern here.

Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/status/protobuf/JobDataWrapperSerializer.scala:83]: +1, we should consider can be repeatedly accessed and used

**Comment by panbingkun** [comment]
> image
> 
> @panbingkun there are 7 usages in from live entities, while there are 13 usages in protobuf serializer..

Base on rule: `a large number of` `duplicated strings with same value`; after refactor it, there are 18 usages in protobuf serializer, as follow:
image

File: core/src/main/scala/org/apache/spark/status/protobuf/PoolDataSerializer.scala [hunk 2]
```diff
@@ -34,7 +33,7 @@ class PoolDataSerializer extends ProtobufSerDe[PoolData] {
   override def deserialize(bytes: Array[Byte]): PoolData = {
     val poolData = StoreTypes.PoolData.parseFrom(bytes)
     new PoolData(
-      name = weakIntern(poolData.getName),
+      name = poolData.getName,
```
Comment (panbingkun) [inline core/src/main/scala/org/apache/spark/status/protobuf/PoolDataSerializer.scala:36]: it violates the rule: `a large number of`, so remove it.

File: core/src/main/scala/org/apache/spark/status/protobuf/StageDataWrapperSerializer.scala [hunk 3]
```diff
@@ -393,10 +393,8 @@ class StageDataWrapperSerializer extends ProtobufSerDe[StageDataWrapper] {
       getOptional(binary.hasFirstTaskLaunchedTime, () => new Date(binary.getFirstTaskLaunchedTime))
     val completionTime =
       getOptional(binary.hasCompletionTime, () => new Date(binary.getCompletionTime))
-    val failureReason =
-      getOptional(binary.hasFailureReason, () => weakIntern(binary.getFailureReason))
-    val description =
-      getOptional(binary.hasDescription, () => weakIntern(binary.getDescription))
+    val failureReason = getOptional(binary.hasFailureReason, binary.getFailureReason)
```
Comment (panbingkun) [inline core/src/main/scala/org/apache/spark/status/protobuf/StageDataWrapperSerializer.scala:396]: ditto

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
Thanks, merging to master

**Comment by panbingkun** [comment]
Thanks @gengliangwang

**Pull request closed without merging**